### DISCUSSION TITLE
Best pattern for parameterising a callback service?

### DISCUSSION THREAD

**Discussion by AGBrown** [statelyai/xstate#1658] [Q&A]
Thank you for putting so much effort into xstate, it is fascinating. I'm using it in a greenfield project and hoping to expand that into older projects as well if it fits well.

I have a use case that seems to fit callback services nicely. The code is small and easily generalisable for the statechart and is used in multiple states. However in each state the code is parameterised slightly differently. A good example would be a merge-event callback where at least one event must be happen on the state, and a minimum delay must also expire since state entry, before the next transition is allowed.

Is there a way to parameterise a callback `service`, or is there a more idiomatic pattern to use? The first thing I reached for was `invoke.data` but it doesn't work for callback services - is that right?

Options:

1. Creating a factory for callback services outside the statechart (e.g. `(delay: number) => (ctx, evt) => (callback, onReceive) => () => void`): the downside is this can't be called by using a key on the `services` item in the machine `options`.
2. Use `entry` to send an event to the callback service, with the callback service "parameters" as the event data: the downside is that gets complicated quickly as you have to wait for the parameters to arrive.
3. Use a child machine with `invoke.data`: downside is this seems like overkill when a small callback would do the job
4. Create a number of `services` (e.g. using the callback service factory from (1)) at machine-declaration time and place them on `options.services` with each under their own key: downside is it bloats `options.services` unnecessarily.
5. Create a parallel state - one to be done after the minimum delay, one to be done after the event. This isn't really reusable other than via copy/paste and doesn't scale well.

For example: https://codesandbox.io/s/xstate-merge-all-s5o98

There are two options in here implemented as parallel statecharts one is a callback function (easy to extract, and so mock, etc.) the other is using parallel states. You can start each demo by trigger the relevant DEMO_CB or DEMO_PSTATE. You won't get to the "finished" stated until the minimum delay has expired && the TRIGGER event is sent to the machine.

I also note that a callback service implemented like this has the weakness that the "RELEASED" event can be triggered externally, not just by the callback. I'm not sure that smells right either.

**Comment by Andarist** [comment]
It looks like parametrized services might be what you are looking for: https://xstate.js.org/docs/guides/communication.html#configuring-services

**Reply by AGBrown** [reply]
👍 I had read about that but not linked it to this use-case. Thank you, I'll try that.
